### ISSUE TITLE
Embedding layer doesn't respect freeze parameter setting

### ISSUE THREAD

**Issue by jason-fries** [HazyResearch/metal#95]
Setting `freeze=True` when initializing the `LSTMModule` should freeze the weight params of the `nn.Embedding` layer object. However, embedding weights seem to update regardless of this setting's value. 

The following code randomly initializes and freezes embedding weights.

```
lstm = LSTMModule(embed_size=50, 
                  hidden_size=100, 
                  vocab_size=train.word_dict.len(), 
                  embeddings=None, 
                  lstm_reduction='max', 
                  dropout=0.0, 
                  num_layers=1, 
                  freeze=True)

end_model = EndModel([200, 2], input_module=lstm, seed=123, use_cuda=False)

# print some embedding values
print(lstm.embeddings(torch.LongTensor([284])))

# train
end_model.train_model(train, dev_data=dev)

# print updated embedding values
print(lstm.embeddings(torch.LongTensor([284])))
```
The code snippet above generates this example output:

```
tensor([[ 0.0884, -2.0748, -0.9850,  0.7471, -0.7564, -0.6341,  0.1372, -0.1688,
          0.7940,  1.0784, -0.8383,  0.1061,  0.8362, -0.7522,  0.5439,  0.2746,
          0.2073, -0.9556,  0.6508, -1.0279, -0.9870,  0.8442, -0.3556, -0.2821,
         -0.4983,  2.3405,  1.8272,  0.9220, -0.9168, -0.4950,  1.4369,  1.0831,
          0.0357, -0.3696, -1.8521,  0.5179, -0.5741,  0.2665,  1.7202,  0.6760,
          1.0778,  1.9969, -0.0641, -1.1610, -0.8505,  0.1784,  2.1454,  1.0016,
         -0.5522,  0.4192]])

tensor([[-0.5792, -0.2044,  0.0411,  1.7785,  0.2299, -0.1451, -1.0534,  0.0911,
         -0.2252, -0.0150,  0.5948,  0.2380, -0.1490,  1.3786,  0.4968,  0.6150,
          2.0698, -1.7208,  0.0183,  0.0287,  1.8081,  0.3915, -0.1944,  0.5217,
          1.4828, -1.3012,  0.1629,  0.8666,  0.6013, -0.6721, -1.2223, -0.0163,
          1.7144, -1.2808,  0.2188, -0.2372, -1.6231, -0.7394, -1.1450,  0.3411,
          0.7713,  0.6864, -0.5711,  1.4876, -1.8473, -0.5141, -0.3369,  0.8294,
          0.4968, -0.6156]])

```
Setting L2 regularization (via `end_model.implements_l2`) doesn't change anything. Manually setting `embeddings.weight.requires_grad` doesn't have any effect. Any ideas why this is happening? These should generate identical values.

**Comment by bhancock8** [comment]
I thought I found the answer (we didn't have a filter in place removing the parameters with requires_grad=False before passing them to the optimizer), but they still seem to be changing. I just added the branch issue95 where I made the change and added a unit test (which is still failing). I have to set this aside for a moment, but it's pushed if either of you want to take a look.

**Comment by bhancock8** [comment]
Found the culprit. The EndModel was getting reset before training, and that reset call was recursive. When an Embeddings object is reset, it gets re-randomly initialized. So it wasn't being updated, it had no gradient, but we were blasting it away when we called train().
